### ISSUE TITLE
Access to undefined reference on nightly

### ISSUE THREAD

**Issue by ararslan** [simonster/JLD2.jl#21]
From the macOS nightly Travis log:
```
ERROR: LoadError: LoadError: UndefRefError: access to undefined reference
Stacktrace:
 [1] iseq(::MyStruct, ::MyStruct) at /Users/travis/.julia/v0.7/JLD2/test/rw.jl:288
 [2] macro expansion at /Users/travis/.julia/v0.7/JLD2/test/rw.jl:318 [inlined]
 [3] macro expansion at ./util.jl:238 [inlined]
 [4] macro expansion at /Users/travis/.julia/v0.7/JLD2/test/rw.jl:475 [inlined]
 [5] anonymous at ./<missing>:?
 [6] include_relative(::Module, ::String) at ./loading.jl:464
 [7] include at ./sysimg.jl:14 [inlined]
 [8] include(::String) at ./client.jl:381
 [9] include_relative(::Module, ::String) at ./loading.jl:464
 [10] include(::Module, ::String) at ./sysimg.jl:14
 [11] process_options(::Base.JLOptions) at ./client.jl:315
 [12] _start() at ./client.jl:383
while loading /Users/travis/.julia/v0.7/JLD2/test/rw.jl, in expression starting on line 359
while loading /Users/travis/.julia/v0.7/JLD2/test/runtests.jl, in expression starting on line 6
```
[Full macOS log](https://travis-ci.org/simonster/JLD2.jl/jobs/256446636). The same happens on [AppVeyor](https://ci.appveyor.com/project/simonster/jld2-jl/build/1.0.37/job/k0xtcr7b028e40ch).

Something slightly different happens on Linux and for whatever reason Travis still reports success:
```
Internal error: encountered unexpected error in runtime:
ErrorException("type does not have a definite number of fields")
rec_backtrace at /home/centos/buildbot/slave/package_tarball64/build/src/stackwalk.c:86
record_backtrace at /home/centos/buildbot/slave/package_tarball64/build/src/task.c:246
jl_throw at /home/centos/buildbot/slave/package_tarball64/build/src/task.c:566
fieldcount at ./reflection.jl:457 [inlined]
fieldnames at ./reflection.jl:143
jl_call_fptr_internal at /home/centos/buildbot/slave/package_tarball64/build/src/julia_internal.h:366 [inlined]
jl_call_method_internal at /home/centos/buildbot/slave/package_tarball64/build/src/julia_internal.h:385 [inlined]
jl_apply_generic at /home/centos/buildbot/slave/package_tarball64/build/src/gf.c:1929
is_allocation at ./inference.jl:5645
...
```
[Full Linux log](https://travis-ci.org/simonster/JLD2.jl/jobs/256446634)

**Comment by simonster** [comment]
Caused by JuliaLang/julia#22929.

**Comment by simonster** [comment]
The `access to undefined reference` error is fixed now, and the tests now "pass" on master, but the `type does not have a definite number of fields` error is still there. I believe this is thrown in inference, so inference just gets disabled and the code still runs. Inference isn't supposed to vomit, so this is probably another upstream bug, but I won't be able to look more closely until the weekend.

**Comment by simonster** [comment]
The warning appears only when code coverage is enabled 😕

**Comment by simonster** [comment]
The error/warning is fixed by #24. We were generating code of "ambiguous validity"; it compiled properly, but the Julia frontend would never produce it, and type inference didn't like it.
